app/services/file_version_membership_diff.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+# @abstract Returns a hash with the differences between the file memberships of two WorkVersions. Renamed files
+# is array of tuples, each with a FileVersionMembership, representating the two changed memberships. Added and deleted
+# files are returned as arrays of FileVersionMembership objects.
+# @example Given two objects with two different titles
+#   >  FileVersionMembershipDiff.call(obj1, obj2)
+#   => {
+#        renamed:
+#          [
+#            [FileVersionMembership, FileVersionMembership],
+#            [FileVersionMembership, FileVersionMembership]
+#          ]
+#        added: [FileVersionMembership, FileVersionMembership],
+#        deleted: [FileVersionMembership]
+#      }
+#
+class FileVersionMembershipDiff
+  # @param obj1 [WorkVersion]
+  # @param obj2 [WorkVersion]
+  def self.call(*args)
+    new(*args).diff
+  end
+
+  attr_reader :base_version, :comparison_version
+
+  def initialize(*args)
+    @base_version = args[0]
+    @comparison_version = args[1]
+  end
+
+  def diff
+    { renamed: renamed, deleted: deleted, added: added }
+  end
+
+  # @return [Array<[FileVersionMembership, FileVersionMembership]>]
+  def renamed
+    renamed_file_resource_ids.map do |id|
+      original_file = base_version_memberships.where(file_resource_id: id).first
+      renamed_file = comparison_version_memberships.where(file_resource_id: id).first
+      next if original_file.title == renamed_file.title
+
+      [original_file, renamed_file]
+    end.compact
+  end
+
+  # @return [Array<FileVersionMembership>]
+  def deleted
+    base_version_memberships.reject do |file_version_membership|
+      renamed_file_resource_ids.include?(file_version_membership.file_resource_id)
+    end
+  end
+
+  # @return [Array<FileVersionMembership>]
+  def added
+    comparison_version_memberships.reject do |file_version_membership|
+      renamed_file_resource_ids.include?(file_version_membership.file_resource_id)
+    end
+  end
+
+  private
+
+    def base_version_memberships
+      @base_version_memberships ||= base_version.file_version_memberships
+    end
+
+    def comparison_version_memberships
+      @comparison_version_memberships ||= comparison_version.file_version_memberships
+    end
+
+    def renamed_file_resource_ids
+      @renamed_file_resource_ids ||= base_version.file_resources.map(&:id) & comparison_version.file_resources.map(&:id)
```

Comment:
This line queries the database for all the `file_resources`, then loads their ActiveRecords into memory, just to pull out their `id`. Since the `FileResourceMembership` contains a foreign key to the FileResource, we can change this to just mapping the `:file_resource_id` on the `file_version_memberships`, this avoiding two database queries. 

Also storing the result of this intersection as some kind of indexed enumerable (Hash or Set) rather than an Array will give us a totally negligible speed boost when looking up the membership in `#added` and `#deleted` since we'd be hitting an index (O(log(n))) rather than iterating through an array (O(n)). Doesn't actually matter for lists this small but will make computer scientists smile.